### PR TITLE
`telepresence login`: Add an `--apikey` flag

### PR DESCRIPTION
**Pull request by LukeShu**
## Description

This PR implements the `telepresence login --apikey` flag which allows specifying a root API key to use instead of performing an interactive login.

## Testing

```console
$ ##############################################################################

$ # 0. Setup

$ ##############################################################################

$ export SYSTEMA_ENV=staging

$ export KUBECONFIG=~/.kube/config.kluster.lukeshu-st.yaml

$ telepresence status
Root Daemon: Not running
User Daemon: Not running

$ # observe that user-info.json does not exist, and that apikeys.json is empty

$ ls ~/.cache/telepresence/
apikeys.json
ingresses.json
logs
update-checks.json

$ jq < ~/.cache/telepresence/apikeys.json
{
  "auth.datawire.io": {},
  "beta-auth.datawire.io": {}
}

$ ##############################################################################

$ # 1. Check that old-style login works

$ ##############################################################################

$ telepresence login
Launching Telepresence User Daemon
Launching browser authentication flow...
Login successful.

$ telepresence status
Root Daemon: Not running
User Daemon: Running
  Version         : v2.3.7-63-g17f1ea07-1626996950 (api 3)
  Ambassador Cloud: Logged in
  Status          : Not connected

$ # check thaat user-info.json exists, and that it was able to exchange the token for a key

$ ls ~/.cache/telepresence/
apikeys.json
ingresses.json
logs
tokens.json
update-checks.json
user-info.json

$ jq < ~/.cache/telepresence/apikeys.json
{
  "auth.datawire.io": {},
  "beta-auth.datawire.io": {
    "telepresence:traffic-manager": "YT…cQ=="
  }
}

$ # Check that logout can undo that 

$ telepresence logout

$ ls ~/.cache/telepresence/
apikeys.json
ingresses.json
logs
update-checks.json

$ jq < ~/.cache/telepresence/apikeys.json
{
  "auth.datawire.io": {},
  "beta-auth.datawire.io": {}
}

$ ##############################################################################

$ # 2. Check that we can change from an access-token login to an apikey login

$ ##############################################################################

$ telepresence login
Launching browser authentication flow...
Login successful.

$ telepresence login --apikey=NG…cA==

$ jq < ~/.cache/telepresence/apikeys.json
{
  "auth.datawire.io": {},
  "beta-auth.datawire.io": {
    "": "NG…cA=="
  }
}

$ # check that it can exchange that key for other keys

$ telepresence connect
Launching Telepresence Root Daemon
Launching Telepresence User Daemon
Connected to context default (https://34.71.233.102)

$ jq < ~/.cache/telepresence/apikeys.json{
  "auth.datawire.io": {},
  "beta-auth.datawire.io": {
    "": "NG…cA==",
    "telepresence:traffic-manager": "ND…bA=="
  }
}

$ telepresence logout

$ jq < ~/.cache/telepresence/apikeys.json{
  "auth.datawire.io": {},
  "beta-auth.datawire.io": {}
}

$ ls ~/.cache/telepresence/
apikeys.json
ingresses.json
logs
update-checks.json

$ ##############################################################################

$ # 3. Check that we can start with apikey

$ ##############################################################################

$ telepresence login --apikey=NG…cA==

$ ls ~/.cache/telepresence/
apikeys.json
ingresses.json
logs
update-checks.json
user-info.json

$ jq < ~/.cache/telepresence/apikeys.json{
  "auth.datawire.io": {},
  "beta-auth.datawire.io": {
    "": "NGNiNWU1MTQtYmY1ZC00YjFhLTllMWYtMDIyZDVmZWFlZTBmOnR6MzJIVG1ibzNkZ0pabFVnaHVzekJhTzREaVFzR3NDeThLcA==",
    "telepresence:traffic-manager": "YWYxYWI3MGYtM2Y3OS00ODE5LWJiNWQtOGM2NzE4ZjQ0NjdkOlNhY2lWZUk5NHF4N0xSSE1NV1ZCV2xYSHlaY0tRajhRdjdxeA=="
  }
}

$ ##############################################################################

$ # 4. Check that bogus keys don't kill live logins

$ ##############################################################################

$ telepresence login --apikey=bogus
telepresence: error: rpc error: code = PermissionDenied desc = status 401 from user info endpoint: permission denied

$ jq < ~/.cache/telepresence/apikeys.json{
  "auth.datawire.io": {},
  "beta-auth.datawire.io": {
    "": "NGNiNWU1MTQtYmY1ZC00YjFhLTllMWYtMDIyZDVmZWFlZTBmOnR6MzJIVG1ibzNkZ0pabFVnaHVzekJhTzREaVFzR3NDeThLcA==",
    "telepresence:traffic-manager": "YWYxYWI3MGYtM2Y3OS00ODE5LWJiNWQtOGM2NzE4ZjQ0NjdkOlNhY2lWZUk5NHF4N0xSSE1NV1ZCV2xYSHlaY0tRajhRdjdxeA=="
  }
}

$ ##############################################################################

$ # 5. Check that logout reports errors

$ ##############################################################################

$ telepresence logout

$ telepresence logout
telepresence: error: Logout: not logged in

$ 
```

## Checklist

 - [x] I made sure to update `./CHANGELOG.md`. - yep
 - [x] I made sure to either submit a docs PR, or tell Matt about the necessary documentation changes. - https://github.com/telepresenceio/telepresence/pull/1876
 - [x] My change is adequately tested. - manually, yes, but automated tests are lacking
 - [x] I updated `DEVELOPING.md` with any any special dev tricks I had to use to work on this code efficiently. - no tricks

### CODE REVIEW

File: pkg/client/connector/userd_trafficmgr/traffic_manager.go [hunk 1]
```diff
@@ -172,7 +173,10 @@ func (tm *trafficManager) Run(c context.Context) error {
 		InstallId: tm.installID,
 		Product:   "telepresence",
 		Version:   client.Version(),
-		ApiKey:    func() string { tok, _ := tm.callbacks.GetCloudAPIKey(c, "manager", false); return tok }(),
+		ApiKey: func() string {
+			tok, _ := tm.callbacks.GetCloudAPIKey(c, a8rcloud.KeyDescTrafficManager, false)
```
Comment (thallgren) [inline pkg/client/connector/userd_trafficmgr/traffic_manager.go:179]: Not that it changes in this PR, but ignoring the error from `GetCloudAPIKey()` seems wrong.

File: pkg/client/connector/userd_trafficmgr/traffic_manager.go [hunk 2]
```diff
@@ -377,7 +381,10 @@ func (tm *trafficManager) remain(c context.Context) error {
 		case <-ticker.C:
 			_, err := tm.managerClient.Remain(c, &manager.RemainRequest{
 				Session: tm.session(),
-				ApiKey:  func() string { tok, _ := tm.callbacks.GetCloudAPIKey(c, "manager", false); return tok }(),
+				ApiKey: func() string {
+					tok, _ := tm.callbacks.GetCloudAPIKey(c, a8rcloud.KeyDescTrafficManager, false)
```
Comment (thallgren) [inline pkg/client/connector/userd_trafficmgr/traffic_manager.go:389]: same here. Why is that error ignored?

File: pkg/client/cli/cliutil/systema.go [hunk 3]
```diff
@@ -59,23 +59,24 @@ func EnsureLoggedOut(ctx context.Context) error {
 // HasLoggedIn returns true if either the user has an active login session or an expired login
 // session, and returns false if either the user has never logged in or has explicitly logged out.
 func HasLoggedIn(ctx context.Context) bool {
-	token, _ := authdata.LoadTokenFromUserCache(ctx)
-	return token != nil
+	userInfo, _ := authdata.LoadUserInfoFromUserCache(ctx)
```
Comment (thallgren) [inline pkg/client/cli/cliutil/systema.go]: Do we want this error to go undetected?

File: pkg/client/cli/cmd_status.go [hunk 4]
```diff
@@ -113,7 +113,7 @@ func connectorStatus(cmd *cobra.Command) error {
 
 		if !cliutil.HasLoggedIn(ctx) {
 			fields = append(fields, kv{"Ambassador Cloud", "Logged out"})
-		} else if _, err := cliutil.GetCloudAccessToken(ctx, false); err != nil {
+		} else if _, err := cliutil.GetCloudUserInfo(ctx, false, true); err != nil {
 			fields = append(fields, kv{"Ambassador Cloud", "Login expired"})
```
Comment (thallgren) [inline pkg/client/cli/cmd_status.go]: This assumes that the only error returned from the `GetCloudUserInfo()` is that the login expired. Other errors will go undetected.

File: pkg/client/connector/userd_grpc/service.go [hunk 5]
```diff
@@ -171,26 +171,18 @@ func (s *service) Logout(ctx context.Context, _ *empty.Empty) (*empty.Empty, err
...
-func (s *service) getCloudAccessToken(ctx context.Context, autoLogin bool) (string, error) {
-	token, err := s.sharedState.LoginExecutor.GetToken(ctx)
-	if autoLogin && err != nil {
+func (s *service) GetCloudUserInfo(ctx context.Context, req *rpc.UserInfoRequest) (*rpc.UserInfo, error) {
+	ctx = s.callCtx(ctx, "GetCloudUserInfo")
+	info, err := s.sharedState.LoginExecutor.GetUserInfo(ctx, req.GetRefresh())
+	if req.GetAutoLogin() && err != nil {
 		if _err := s.sharedState.LoginExecutor.Login(ctx); _err == nil {
-			token, err = s.sharedState.LoginExecutor.GetToken(ctx)
+			info, err = s.sharedState.LoginExecutor.GetUserInfo(ctx, req.GetRefresh())
```
Comment (thallgren) [inline pkg/client/connector/userd_grpc/service.go]: This means that all kinds of errors that the `LoginExecutor.Login()` can return are lost. Is that intentional?

**Review by thallgren — APPROVED** [review]
Looks good. I made a few comments regarding bad error handling. Not introduced in this PR but that it might be good to fix while touching that code. Not a blocker though.

**Comment by thallgren** [comment]
Will you copy that testing walkthrough to the manual smoke test document?

File: pkg/client/cli/cliutil/systema.go (hunk 3 above)
Comment (LukeShu) [reply pkg/client/cli/cliutil/systema.go]: We do.  On the invariant that `(userInfo == nil) != (err == nil)`, we only need to look at one.  Perhaps I should have used `_, err := …; return err == nil`.

File: pkg/client/cli/cmd_status.go (hunk 4 above)
Comment (LukeShu) [reply pkg/client/cli/cmd_status.go]: That's pretty dang close to being true.  "We've previously logged in, but the login is no longer functioning"

File: pkg/client/connector/userd_grpc/service.go (hunk 5 above)
Comment (LukeShu) [reply pkg/client/connector/userd_grpc/service.go]: Indeed it is.

File: pkg/client/connector/userd_trafficmgr/traffic_manager.go (hunk 2 above)
Comment (LukeShu) [reply pkg/client/connector/userd_trafficmgr/traffic_manager.go:389]: Including an apikey with this request is optional; we might not even be logged in.

File: pkg/client/connector/userd_trafficmgr/traffic_manager.go (hunk 1 above)
Comment (LukeShu) [reply pkg/client/connector/userd_trafficmgr/traffic_manager.go:179]: Including an apikey with this request is optional; we might not even be logged in.

**Comment by LukeShu** [comment]
Rebased to resolve conflicts in `CHANGELOG.md`.  Added a commit at the end improving clarity around error handling.

**Pull request merged**